blockstore/splitstore/splitstore_prune.go

```diff
@@ -0,0 +1,592 @@
+	//   this is useful for garbage collecting side-chains and other garbage in archival nodes.
+	//   This is the (safe) default.
+	// - if it is 0 then no state objects that are unreachable within the compaction boundary will
+	//   be retained in the coldstore.
+	// - if it is a positive integer, then it's the number of finalities past the compaction boundary
+	//   for which chain-reachable state objects are retained.
+	PruneRetainState = "splitstore.PruneRetainState"
+)
+
+// PruneChain instructs the SplitStore to prune chain state in the coldstore, according to the
+// options specified.
+func (s *SplitStore) PruneChain(opts map[string]interface{}) error {
+	// options
+	var onlineGC, movingGC bool
+	var retainState int64 = -1
+
+	for k, v := range opts {
+		switch k {
+		case PruneOnlineGC:
+			onlineGC = true
+		case PruneRetainState:
+			retaini64, ok := v.(int64)
+			if !ok {
+				// deal with json-rpc types...
+				retainf64, ok := v.(float64)
+				if !ok {
+					return xerrors.Errorf("bad state retention specification; expected int64 or float64 but got %T", v)
+				}
+				retainState = int64(retainf64)
+			} else {
+				retainState = retaini64
+			}
+		default:
+			return xerrors.Errorf("unrecognized option %s", k)
+		}
+	}
+
+	if onlineGC && movingGC {
+		return xerrors.Errorf("at most one of online, moving GC can be specified")
+	}
+	if !onlineGC && !movingGC {
+		onlineGC = true
+	}
+	var gcOpts []bstore.BlockstoreGCOption
+	if movingGC {
+		gcOpts = append(gcOpts, bstore.WithFullGC(true))
+	}
+	doGC := func() error { return s.gcBlockstore(s.cold, gcOpts) }
+
+	var retainStateP func(int64) bool
+	switch {
+	case retainState > 0:
+		retainStateP = func(depth int64) bool {
+			return depth <= int64(CompactionBoundary)+retainState*int64(build.Finality)
+		}
+	case retainState < 0:
+		retainStateP = func(_ int64) bool { return true }
+	default:
+		retainStateP = func(depth int64) bool {
+			return depth <= int64(CompactionBoundary)
+		}
+	}
+
+	if _, ok := s.cold.(bstore.BlockstoreIterator); !ok {
+		return xerrors.Errorf("coldstore does not support efficient iteration")
+	}
+
+	return s.pruneChain(retainStateP, doGC)
+}
+
+func (s *SplitStore) pruneChain(retainStateP func(int64) bool, doGC func() error) error {
+	// inhibit compaction while we are setting up
+	s.headChangeMx.Lock()
+	defer s.headChangeMx.Unlock()
+
+	// take the compaction lock; fail if there is a compaction in progress
+	if !atomic.CompareAndSwapInt32(&s.compacting, 0, 1) {
+		return xerrors.Errorf("compaction, prune or warmup in progress")
+	}
+
+	// check if we are actually closing first
+	if atomic.LoadInt32(&s.closing) == 1 {
+		atomic.StoreInt32(&s.compacting, 0)
+		return errClosing
+	}
+
+	// ensure that we have compacted at least once
+	if s.compactionIndex == 0 {
+		atomic.StoreInt32(&s.compacting, 0)
+		return xerrors.Errorf("splitstore has not compacted yet")
+	}
+
+	// get the current tipset
+	curTs := s.chain.GetHeaviestTipSet()
```

Comment:
Don't we technically want the last tipset we've seen by the head change handler, or does that not really matter?

Reply:
I don't think it matters much.  This code path is hit through a direct call to `chain prune` so I think users expect to prune against whatever the latest head is.